File: backend/app/domain/goal_repository.py
from abc import ABC, abstractmethod
from datetime import date
from uuid import UUID

from app.domain.monthly_goal import MonthlyGoal


class GoalRepository(ABC):
    @abstractmethod
    async def list_for_month(self, user_id: UUID, month: date) -> list[MonthlyGoal]: ...

    @abstractmethod
    async def add(self, goal: MonthlyGoal) -> MonthlyGoal: ...

    @abstractmethod
    async def increment_progress(self, goal_id: UUID) -> MonthlyGoal | None: ...

File: frontend/lib/features/goals/domain/goal_repository.dart
import 'monthly_goal.dart';

class TooManyActiveGoalsError implements Exception {}

abstract class GoalRepository {
  /// Active (not yet complete) goals for the current month, oldest first.
  Future<List<MonthlyGoal>> getActiveGoals();

  /// Throws [TooManyActiveGoalsError] if the user already has
  /// [MonthlyGoal.maxActiveGoals] active goals this month.
  Future<MonthlyGoal> createGoal({required String title, required num target, required String unit});

  /// Adds [amount] units of progress to the goal, capped at its target.
  /// [amount] may be fractional (e.g. 0.25 of an "hours" goal).
  Future<void> incrementProgress(String goalId, {required num amount});

  Future<void> updateGoal({
    required String goalId,
    required String title,
    required num target,
    required String unit,
  });

  Future<void> deleteGoal(String goalId);
}
